internal/pkg/service/stream/sink/router/pipeline.go

```diff
@@ -26,24 +28,80 @@ type pipelineRef struct {
 	openRetryAfter time.Time
 }
 
+func (p *pipelineRef) writeRecord(c recordctx.Context) (pipeline.RecordStatus, error) {
+	if err := p.ensureOpened(c.Ctx(), c.Timestamp()); err != nil {
+		return pipeline.RecordError, err
+	}
+	return p.pipeline.WriteRecord(c)
+}
+
```

Comment:
Offtopic: Shuffled the code so that it is clear how it works, and changes can be made there (next PR).

Reply:
Here you can see that the pipeline is lazy, and it is opened at the first use.